lib/nolibc, lib/isrlib: Pointer comparison

ranfun opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.pl`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.pl) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->
The memmove_isr function is used to copy a block of memory from a source address to a destination address. The function is used in interrupt service routines (ISRs) where it is important to copy the memory as quickly as possible.

In the previous implementation of memmove_isr, the function assumed that the source address is either greater than the destination address or less than it. If both the conditions were false, the function would not work as expected and may cause undefined behavior. This assumption could lead to issues where the memory blocks overlap.

The modified implementation of memmove_isr takes into account both overlapping and non-overlapping cases. The function typecasts the src and dst pointers to intptr_t to ensure that the values being compared are treated as pointers rather than integers. This can help prevent errors that might occur if the pointer values are interpreted as signed integers.  ( Issue [#447](https://github.com/unikraft/unikraft/issues/447) ).

The changes were implemented in both nolibc and isrlib libraries to ensure that the code is consistent across all libraries and to prevent any issues when switching between them.

Overall, the changes ensure that memmove_isr works correctly in all cases, including when the memory blocks overlap, and that the code is robust and less prone to errors.

ranfun committed:
lib/nolibc, lib/isrlib: Fixed warnings for on-pointer comparison (memmove)

Typecasted to ensure values compared are treated as pointers and not integers
Modified implementation to handle both overlapping and non-overlapping cases.

Signed-off-by: Soumya Ranjan <ranjan42069@gmail.com>

ranfun committed:
lib/nolibc, lib/isrlib: Fixed warnings for on-pointer comparison (memmove)

Typecasted to ensure values compared are treated as pointers and not integers
Modified implementation to handle both overlapping and non-overlapping cases.

Signed-off-by: Soumya Ranjan <ranjan42069@gmail.com>

lib/{nolibc,isrlib}:modify pointer comparison

typecast to ensure values treated as pointers
modify implementation to handle overlapping

Signed-off-by: Soumya Ranjan <ranjan42069@gmail.com>

ranfun committed:
lib/nolibc, lib/isrlib: Modified pointer comparison (memmove)

Typecasted to ensure values treated as pointers and not integers
Modified implementation to handle overlapping and non-overlapping cases.

Signed-off-by: Soumya Ranjan <ranjan42069@gmail.com>

ranfun committed:
lib/nolibc, lib/isrlib: Modified pointer comparison (memmove)

Typecasted to ensure values treated as pointers and not integers
Modified implementation to handle overlapping and non-overlapping cases.

Signed-off-by: Soumya Ranjan <ranjan42069@gmail.com>

ranfun committed:
lib/nolibc, lib/isrlib: Modified pointer comparison (memmove)

Typecasted to ensure values treated as pointers and not integers
Modified implementation to handle overlapping and non-overlapping cases.

Signed-off-by: Soumya Ranjan <ranjan42069@gmail.com>

mariasfiraiala commented on lib/isrlib/string.c:
```diff
@@ -108,18 +108,22 @@ void *memrchr_isr(const void *m, int c, size_t n)
-	}
+    const uint8_t *s = src;
+
+    if ((intptr_t)src == (intptr_t)dst) {
+        return dst;
+    }
```
```suggestion
    if ((intptr_t)src == (intptr_t)dst)
        return dst;
```

unnecessary braces

mariasfiraiala commented on lib/nolibc/string.c:
```diff
@@ -109,7 +109,11 @@ void *memmove(void *dst, const void *src, size_t len)
 	const uint8_t *s = src;
 
-	if (src > dst) {
+	if ((intptr_t)src == (intptr_t)dst) {
+        return dst;
+    }
```
Suspect code indent and unnecessary braces

mariasfiraiala reviewed (changes requested):
Thanks @ranfun, the changes look ok. However, a few cosmetic aspects should be taken into consideration before merging this:

* Run `checkpatch.pl` on your commits using the instructions from [here](https://unikraft.org/docs/contributing/testing/#using-checkpatchpl) and fix your coding style errors. You'll find a couple of examples in my comments below.

* Squash your commits. The 4 of them can be combined into a single one, which suffices.

* Modify your commit message such as the verbs are all at Present Tense. Use the prefix `lib/{nolibc,isrlib}` before the description of the changes in your commit message. A good practice is to also have less than 50 characters for the commit message, so try to accomplish that.

eduardvintila reviewed (commented):
Hey, @ranfun. Thanks for the changes! Please follow @mariasfiraiala's suggestions so we can merge this PR : )

ranfun commented:
yes will do!


On Fri, Apr 21, 2023 at 1:03 PM Eduard Vintilă ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
>
> Hey, @ranfun <https://github.com/ranfun>. Thanks for the changes! Please
> follow @mariasfiraiala <https://github.com/mariasfiraiala>'s suggestions
> so we can merge this PR : )
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/unikraft/unikraft/pull/825#pullrequestreview-1396338355>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ARPASJYYZFXNMDLTT7XRJQ3XCLR27ANCNFSM6AAAAAAWQR3Y5U>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

razvand commented:
@ranfun, can you take care of this, please?

ranfun committed:
lib/{nolibc,isrlib}:Modify pointer comparison

Typecast to ensure value treated as pointer
Modify implementation to handle overlapping

Signed-off-by: Soumya Ranjan <ranjan42069@gmail.com>

razvand commented:
@ranfun, please see the checkpatch warnings

ranfun commented:
The latest patch passed the checkpatch warnings

mariasfiraiala commented:
> The latest patch passed the checkpatch warnings

This is exactly why you should drop the other 5 commits.  We shouldn't have upstream unnecessary patches, such as the first 5 commits that **do not** pass the coding style checker.

So, @ranfun, drop them.

mariasfiraiala commented:
Closed via #880.
